beacon-chain/db/kv/blobs.go

```diff
@@ -71,91 +106,76 @@ func (s *Store) BlobsSidecar(ctx context.Context, blockRoot [32]byte) (*ethpb.Bl
+		blob := &ethpb.BlobsSidecar{}
+		if err := decode(ctx, enc, blob); err != nil {
+			return nil, err
+		}
+		blobs[i] = blob
+	}
+	return blobs, nil
 }
 
 // HasBlobsSidecar returns true if the blobs are in the db.
-func (s *Store) HasBlobsSidecar(ctx context.Context, root [32]byte) bool {
+func (s *Store) HasBlobsSidecar(ctx context.Context, beaconBlockRoot [32]byte) bool {
 	ctx, span := trace.StartSpan(ctx, "BeaconDB.HasBlobsSidecar")
 	defer span.End()
-
-	exists := false
-	if err := s.db.View(func(tx *bolt.Tx) error {
-		bkt := tx.Bucket(blobsBucket)
-		exists = bkt.Get(root[:]) != nil
-		return nil
-	}); err != nil { // This view never returns an error, but we'll handle anyway for sanity.
-		panic(err)
+	blobSidecar, err := s.BlobsSidecar(ctx, beaconBlockRoot)
+	if err != nil {
+		return false
 	}
-	return exists
+	return blobSidecar != nil
 }
 
-// CleanupBlobs removes blobs that are older than the cutoff time.
-func (s *Store) CleanupBlobs(ctx context.Context) error {
-	ctx, span := trace.StartSpan(ctx, "BeaconDB.CleanupBlobs")
+// DeleteBlobsSidecar returns true if the blobs are in the db.
+func (s *Store) DeleteBlobsSidecar(ctx context.Context, beaconBlockRoot [32]byte) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.DeleteBlobsSidecar")
 	defer span.End()
-
-	secsInEpoch := time.Duration(params.BeaconConfig().SlotsPerEpoch.Mul(params.BeaconConfig().SecondsPerSlot)) * time.Second
-	ttl := secsInEpoch * (time.Duration(params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest) + 1) // add one more epoch as slack
-
-	var expiredBlobs [][]byte
-	now := time.Now()
-	err := s.db.View(func(tx *bolt.Tx) error {
-		bkt := tx.Bucket(blobsAgesBucket)
-		c := bkt.Cursor()
-		for k, v := c.First(); k != nil; k, v = c.Next() {
-			insertTime, err := time.Parse(time.RFC3339, string(v))
-			if err != nil {
-				return err
-			}
-			if now.Sub(insertTime) > ttl {
-				expiredBlobs = append(expiredBlobs, k)
-			}
-		}
-		return nil
-	})
-	if err != nil {
-		return err
-	}
-
-	log.WithField("count", len(expiredBlobs)).Info("Cleaning up blobs")
-
 	return s.db.Update(func(tx *bolt.Tx) error {
-		agesBkt := tx.Bucket(blobsAgesBucket)
 		bkt := tx.Bucket(blobsBucket)
-		for _, root := range expiredBlobs {
-			if err := bkt.Delete(root); err != nil {
-				return err
-			}
-			if err := agesBkt.Delete(root); err != nil {
-				return err
+		c := bkt.Cursor()
+		for k, _ := c.First(); k != nil; k, _ = c.Next() {
+			if bytes.HasSuffix(k, beaconBlockRoot[:]) {
+				if err := bkt.Delete(k); err != nil {
+					return nil
+				}
 			}
 		}
 		return nil
 	})
 }
+
+// We define a blob sidecar key as: bytes(slot_to_rotating_buffer(blob.slot)) ++ bytes(blob.slot) ++ blob.block_root
+// where slot_to_rotating_buffer(slot) = slot % MAX_SLOTS_TO_PERSIST_BLOBS.
+func blobSidecarKey(blob *ethpb.BlobsSidecar) []byte {
+	slotsPerEpoch := params.BeaconConfig().SlotsPerEpoch
+	maxEpochsToPersistBlobs := params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
```

Comment:
Should this be configurable via a flag? I can imagine users wanting to save blobs for longer.

Reply:
Yep it should be done via a feature config IMO...will reserve that for a second PR